crates/sui-indexer-alt-framework/src/pipeline/concurrent/committer.rs

```diff
@@ -213,3 +213,513 @@ pub(super) fn committer<H: Handler + 'static>(
         }
     })
 }
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::HashMap,
+        sync::{
+            atomic::{AtomicUsize, Ordering},
+            Arc, Mutex,
+        },
+    };
+
+    use async_trait::async_trait;
+    use sui_types::full_checkpoint_content::CheckpointData;
+    use tokio::sync::mpsc;
+    use tokio_util::sync::CancellationToken;
+
+    use crate::{
+        metrics::IndexerMetrics,
+        pipeline::{
+            concurrent::{BatchedRows, Handler},
+            Processor, WatermarkPart,
+        },
+        store::CommitterWatermark,
+        testing::mock_store::*,
+        FieldCount,
+    };
+
+    use super::*;
+
+    #[derive(Clone, FieldCount)]
+    pub struct StoredData {
+        pub cp_sequence_number: u64,
+        pub tx_sequence_numbers: Vec<u64>,
+        /// Tracks remaining commit failures for testing retry logic.
+        /// The committer spawns concurrent tasks that call H::commit,
+        /// so this needs to be thread-safe (hence Arc<AtomicUsize>).
+        pub commit_failure_remaining: Arc<AtomicUsize>,
+        pub commit_delay_ms: u64,
+    }
+
+    impl Default for StoredData {
+        fn default() -> Self {
+            Self {
+                cp_sequence_number: 0,
+                tx_sequence_numbers: Vec::new(),
+                commit_failure_remaining: Arc::new(AtomicUsize::new(0)),
+                commit_delay_ms: 0,
+            }
+        }
+    }
+
+    pub struct DataPipeline;
+
+    impl Processor for DataPipeline {
+        const NAME: &'static str = "data";
+
+        type Value = StoredData;
+
+        fn process(&self, _checkpoint: &Arc<CheckpointData>) -> anyhow::Result<Vec<Self::Value>> {
+            Ok(vec![])
+        }
+    }
+
+    #[async_trait]
+    impl Handler for DataPipeline {
+        type Store = MockStore;
+
+        async fn commit<'a>(
+            values: &[StoredData],
+            conn: &mut MockConnection<'a>,
+        ) -> anyhow::Result<usize> {
+            for value in values {
+                // If there's a delay, sleep for that duration
+                if value.commit_delay_ms > 0 {
+                    tokio::time::sleep(Duration::from_millis(value.commit_delay_ms)).await;
+                }
+
+                // If there are remaining failures, fail the commit and decrement the counter
+                {
+                    let remaining = value
+                        .commit_failure_remaining
+                        .fetch_sub(1, Ordering::SeqCst);
+                    if remaining > 0 {
+                        return Err(anyhow::anyhow!(
+                            "Commit failed, remaining failures: {}",
+                            remaining - 1
+                        ));
+                    }
```

Comment:
I don't think you need sequential consistency here -- relaxed ordering should be fine, because you just need to make sure that the value is self-consistent, which you get because it is Atomic. The Ordering parameter is about the order of this operation relative to operations on other values that can be witnessed across threads, which is not relevant here.

```rust
let remaining = value
    .commit_failure_remaining
    .fetch_sub(1, Ordering::Relaxed);
ensure!(remaining == 0, "Commit failed, remaining failures: {}", remaining - 1);
```